### PR TITLE
Task Supervisor for message consumers

### PR DESCRIPTION
**Pull request by akoutmos**
This PR is an initial attempt at addressing issue #46. Still in the process of adding tests and solidifying the implementation, so leaving the PR in draft until complete (also looking for implementation feedback). 

## Description
Originally, the `GenRMQ.Consumer` module would call `spawn/1` to asynchronously execute a message handler. This change instead leverages `Task.Supervisor` to execute message processing handlers. This should be a transparent change to the end user as this does not impact how `GenRMQ.Consumer` concurrently executes message handlers.

This introduces  the usage of OTP 21's `handle_continue` to ensure that the `Task.Supervisor` is started before any mailbox messages are processed. As a result, gen_rmq would no longer be compatible with OTP releases prior to 21.

## Checklist
- [x] I have added unit tests to cover my changes.
- [x] I have improved the code quality of this repo. (refactoring, or reduced number of static analyser issues)
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,9 +3,9 @@ elixir:
   - 1.7
   - 1.8
   - 1.9
+  - 1.10
 
 otp_release:
-  - 20.0
   - 21.0
```
Comment (mkorszun) [inline .travis.yml:9]: I am fine with dropping support for this version. We will have to release a new major version though.

File: lib/consumer.ex [hunk 2]
```diff
@@ -298,6 +297,20 @@ defmodule GenRMQ.Consumer do
     |> handle_reconnect(state)
   end
 
+  @doc false
+  @impl GenServer
+  def handle_info({:DOWN, _ref, :process, _pid, _reason}, state) do
```
Comment (mkorszun) [inline lib/consumer.ex:302]: Will we get here messages for tasks or supervisor?

**Comment by mkorszun** [comment]
@akoutmos looking good - left some questions.

File: .travis.yml (hunk 1 above)
Comment (spier) [reply .travis.yml:9]: Just curious, what are we winning by dropping support for this version?

Comment (mkorszun) [reply .travis.yml:9]: * we can use new features from otp21
* we can continue upgrading `amqp` library, which right now I believe requires otp21 as well

Comment (akoutmos) [reply .travis.yml:9]: The primary reason for the change was so that `handle_continue` could be used to ensure that the init parts are called before anything else in the GenServer.

Comment (spier) [reply .travis.yml:9]: Got it. Thanks!

File: lib/consumer.ex (hunk 2 above)
Comment (akoutmos) [reply lib/consumer.ex:302]: This should be coming from the task if it failed. I'm thinking about adding a telemetry event in there to capture the error so library users can track processing failures.

File: .travis.yml (hunk 1 above)
Comment (spier) [reply .travis.yml:9]: @mkorszun feel free to resolve this conversation, unless you need it as a reminder to "release a new major version".

**Comment by akoutmos** [comment]
I added some preliminary tests for this functionality, updated the telemetry documentation, and also refactored the consumer a bit to be able to act upon Tasks that it has spawned under the supervisor. There are still a few more edge cases that I want to cover in the tests, so I will keep this in draft until I feel as though I have covered those code paths.

**Comment by akoutmos** [comment]
I think this one is at a good point to review now. I added the ability for consumer processes to wait on in flight tasks under the Task.Supervisor before terminating and also added tests to validate the flow. I also added an additional telemetry event to report on when Tasks fail along with some tests to cover these paths.

One question I have is if we should expose an optional configuration for how long to wait on running Tasks. Currently I am just using the default value that is set for `yield_many` https://hexdocs.pm/elixir/Task.html#yield_many/2

**Comment by vorce** [comment]
@akoutmos I think yes we have to let users chose the task timeout. I suspect users will have very different expectations on their processor, depending on their business logic.

**Comment by akoutmos** [comment]
I'll add that tonight and leave the default as the yield_many value.

**Comment by spier** [comment]
@akoutmos can I just say: You are awesome :) Great contributions in this PR (and elsewhere). Thanks so much!

**Comment by akoutmos** [comment]
Thanks for the kind words @spier! I appreciate it :)

**Comment by akoutmos** [comment]
Went ahead and added the `terminate_timeout` option for users to configure how long to wait on in-flight Tasks. I also updated the docs to let them know about the option.

**Comment by mkorszun** [comment]
@akoutmos amazing job 🥇- I will reserve some time during the weekend to review it.

File: lib/consumer.ex [hunk 3]
```diff
@@ -276,26 +288,42 @@ defmodule GenRMQ.Consumer do
...
-    state =
-      state
-      |> get_connection()
-      |> open_channels()
-      |> setup_consumer()
+  def handle_info(
+        {:DOWN, ref, :process, _pid, reason},
+        %{module: module, config: config, running_tasks: running_tasks} = state
+      ) do
+    if Map.has_key?(running_tasks, ref) do
```
Comment (vorce) [inline lib/consumer.ex:303]: What happens in this case to the message? Will it be rejected/nacked?

File: test/gen_rmq_consumer_test.exs [hunk 4]
```diff
@@ -92,6 +97,80 @@ defmodule GenRMQ.ConsumerTest do
...
+
+    setup do
+      Agent.start_link(fn -> MapSet.new() end, name: ErrorInConsumer)
+      with_test_consumer(ErrorInConsumer)
+    end
+
+    test "should invoke the consumer's handle_info callback if error exists",
+         %{consumer: consumer_pid, state: state} = context do
+      clear_mailbox()
+      message = %{"value" => 0}
```
Comment (vorce) [inline test/gen_rmq_consumer_test.exs:111]: Can we make it clearer how this triggers an error in the consumer? Even just a comment would help clarify that this will trigger a division by zero.
Or make it even more explicit in ErrorConsumer with something like `if message.value == 0, do: raise "Error!"`

File: lib/consumer.ex [hunk 5]
```diff
@@ -59,12 +59,16 @@ defmodule GenRMQ.Consumer do
   `queue_max_priority` - defines if a declared queue should be a priority queue.
   Should be set to a value from `1..255` range. If it is greater than `255`, queue
   max priority will be set to `255`. Values between `1` and `10` are
-  [recommened](https://www.rabbitmq.com/priority.html#resource-usage).
+  [recommended](https://www.rabbitmq.com/priority.html#resource-usage).
```
Comment (mkorszun) [inline lib/consumer.ex:62]: 👍

File: lib/consumer.ex (hunk 3 above)
Comment (akoutmos) [reply lib/consumer.ex:303]: Good call....we would probably want some sort of configuration so the user can decide what happens to the message if the task fails. Perhaps we also provide the option to send the message to the deadletter exchange. Also on that same note, we may also want some configuration around how long the Tasks take to complete. Currently it is set to the `async_nolink` value of 5 seconds. Thoughts?

Comment (vorce) [reply lib/consumer.ex:303]: Maybe you should be able to set a callback to deal with the message? But I think the sensible default here would be to nack the message.

Comment (mkorszun) [reply lib/consumer.ex:303]: I like the callback, smth like: `handle_error(reason, state)`. Then in this callback user has all the power to ack, reject or requeue.

Comment (mkorszun) [reply lib/consumer.ex:303]: btw, are we capturing here also task timeouts?

File: README.md [hunk 6]
```diff
@@ -75,7 +75,7 @@ This will result in:
 - durable `gen_rmq_in_queue_error` queue created or redeclared. It will be bound to `gen_rmq_exchange.deadletter`
 - durable topic `gen_rmq_exchange` exchange created or redeclared
 - durable `gen_rmq_in_queue` queue created or redeclared. It will be bound to `gen_rmq_exchange` exchange and has a deadletter exchange set to `gen_rmq_exchange.deadletter`
-- every `handle_message` callback will executed in separate process. This can be disabled by setting `concurrency: false` in `init` callback
+- every `handle_message` callback will executed in separate supervised Task. This can be disabled by setting `concurrency: false` in `init` callback
```
Comment (mkorszun) [inline README.md:78]: `will executed in separate supervised Task` -> `will be executed in separate supervised Task`

**Comment by akoutmos** [comment]
Another thing that I thought about today was that there is no way to throttle the consumer. With how it currently is, it can keep on consuming messages and spawning new tasks under the supervisor. Thoughts on adding an additional configuration to have the user set the number of concurrent tasks running? `max_concurrent_tasks` ?

**Comment by vorce** [comment]
@akoutmos good point! I think we can introduce that later if it turns out to be a problem for some users. The current/master implementation has the same problem.

File: lib/consumer.ex (hunk 3 above)
Comment (akoutmos) [reply lib/consumer.ex]: My latest commit adds the ability to configure a timeout for tasks. It was a little tricky to add that in given that Task.Supervisor does not directly support it, but I think I have a good implementation in place to handle this functionality.

As for a custom callback to handle the failure, are you thinking about rewriting the consumer module to a macro so that we can leverage `defoverridable` if the user does not want to use the default error handler?

Comment (mkorszun) [reply lib/consumer.ex]: @akoutmos I would prefer to avoid rewriting consumer module to a macro at this stage (just to limit the scope of changes).

Since these changes will be released as a major version, requiring error callback to be implemented should be fine, right? @vorce what do you think?

We could also consider skipping error callback and just reject the messages. The problem here is that users might skip dead-letter configuration for their consumers.

Comment (vorce) [reply lib/consumer.ex]: Hm hm. Yeah I think making each consumer implement a callback is OK. At least then we encourage users to think about this case. We can have a good implementation in the example consumer that logs and nacks the message.

Comment (akoutmos) [reply lib/consumer.ex]: Sounds good. I'll add an additional callback `handle_error` that will be triggered on task exception or task timeout.

Comment (vorce) [reply lib/consumer.ex]: Awesome, and thanks again for all the work @akoutmos

**Comment by akoutmos** [comment]
I added the `handle_error` callback for consumers that are configured to execute via Tasks. I'll need to also hook it up to trigger the `handle_error` callback when running in synchronous mode. But wanted to get some feedback on the approach before adding that in as well.

File: lib/consumer.ex [hunk 7]
```diff
@@ -186,6 +187,28 @@ defmodule GenRMQ.Consumer do
...
   @callback handle_message(message :: %GenRMQ.Message{}) :: :ok
 
+  @doc """
+  Invoked when an error or timeout is encountered while executing `handle_message` callback
+
+  `message` - `GenRMQ.Message` struct
+  `reason` - atom denoting the type of error
+
+  ## Examples:
+  To reject the message message that caused the Task to fail you can do something like so:
```
Comment (mkorszun) [inline lib/consumer.ex:197]: double `message`

File: lib/message_task.ex [hunk 8]
```diff
@@ -0,0 +1,25 @@
+defmodule GenRMQ.MessageTask do
```
Comment (mkorszun) [inline lib/message_task.ex:1]: Good idea to wrap it like this 👍

**Comment by mkorszun** [comment]
@akoutmos error handling looks good 👍

**Comment by akoutmos** [comment]
I wrapped up that last of the tests and added `handle_error` callback support for non concurrent consumers. I'll give this a final look over tomorrow...but I think from a functionality perspective there is nothing else on my todo list. Let me know if you want to see any more tests/features :)

**Comment by mkorszun** [comment]
@akoutmos great, thank you 🙏 I plan to merge it today/tomorrow.

**Comment by akoutmos** [comment]
Updated some docs and some minor error handling tweaks. Nothing else pops out at me for clean up. All good on this side

**Comment by spier** [comment]
🎉

**Comment by spier** [comment]
@mkorszun just curious, what is the release process for the library? I remember you mentioning that we would have to release a new major(?) version for this?

**Comment by mkorszun** [comment]
@spier we need a major release because we have dropped support for some older versions of the OTP. Release process will involve:
* bumping lib version to 3.0.0
* documentation differences / migration steps between v2 and v3
* merging all above changes to master and do a github release
* publish to hex.pm

**Comment by spier** [comment]
@mkorszun Thanks for the writeup.

I am going through the steps, to understand them better. I might write a new issue to document the process for ourselves, and additional admins in the future. I will try to review what you do for this 3.0.0 release and then try myself on writing this up if that is ok?

## Notes

- bumping lib version to 3.0.0 => at least in [mix.exs](https://github.com/meltwater/gen_rmq/blob/master/mix.exs) and `README.md`.
- documenting differences / migration steps between v2 and v3 => probably in the [wiki](https://github.com/meltwater/gen_rmq/wiki/Migrations)? I looked for migrations steps from v1 to v2 as a sample. maybe we did not write those then?
- merging all above changes to master and doing a github release => [releases](https://github.com/meltwater/gen_rmq/releases)
- publishing to hex.pm => https://hex.pm/packages/gen_rmq

## (unrelated) 30k party

Btw I realized that we are at `28173` **downloads all time** and currently at `803` **downloads last 7 days**. So we should hit the 30k downloads milestone in about 2.5 weeks. How does one throw a virtual party? :)

**Comment by mkorszun** [comment]
@spier sounds good. Your notes are already quite precise :)

regarding `30k party` - I believe we need it 🎉
